### PR TITLE
Add ArrayFindFunctionReturnTypeExtension

### PR DESCRIPTION
**Pull request by zonuexe**
Supports [PHP RFC: array_find](https://wiki.php.net/rfc/array_find) for PHP 8.4.

### CODE REVIEW

File: tests/PHPStan/Analyser/nsrt/array-find.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		assertType('int|null', array_find($array, 'is_int'));
+	}
+
+	/**
+	 * @param array{1, 'foo', \DateTime} $array
+	 */
+	function testConstant(array $array, callable $callback): void
+	{
+		assertType("1|'foo'|DateTime|null", array_find($array, $callback));
+		assertType("1|null", array_find($array, 'is_int'));
```
Comment (VincentLanglet) [inline tests/PHPStan/Analyser/nsrt/array-find.php]: This cannot be Null.

File: tests/PHPStan/Analyser/nsrt/array-find.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+	use function PHPStan\Testing\assertType;
+
+	/**
+	 * @param array<mixed> $array
+	 */
+	function testMixed(array $array, callable $callback): void
+	{
+		assertType('mixed', array_find($array, $callback));
+		assertType('int|null', array_find($array, 'is_int'));
```
Comment (VincentLanglet) [inline tests/PHPStan/Analyser/nsrt/array-find.php:38]: You should also test it on non empty array, and the result should be an int

File: tests/PHPStan/Analyser/nsrt/array-find-key.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		assertType('int|string|null', array_find_key($array, 'is_int'));
+	}
+
+	/**
+	 * @param array{1, 'foo', \DateTime} $array
+	 */
+	function testConstant(array $array, callable $callback): void
+	{
+		assertType("0|1|2|null", array_find_key($array, $callback));
+		assertType("0|1|2|null", array_find_key($array, 'is_int'));
```
Comment (VincentLanglet) [inline tests/PHPStan/Analyser/nsrt/array-find-key.php:47]: It should be 0

File: tests/PHPStan/Analyser/nsrt/array-find-key.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+	use function PHPStan\Testing\assertType;
+
+	/**
+	 * @param array<mixed> $array
+	 */
+	function testMixed(array $array, callable $callback): void
+	{
+		assertType('int|string|null', array_find_key($array, $callback));
+		assertType('int|string|null', array_find_key($array, 'is_int'));
```
Comment (VincentLanglet) [inline tests/PHPStan/Analyser/nsrt/array-find-key.php:37]: You should test on non-empty-array<int> and it should gives int|string

File: tests/PHPStan/Analyser/nsrt/array-find-key.php (hunk 3 above)
Comment (zonuexe) [reply tests/PHPStan/Analyser/nsrt/array-find-key.php:47]: Ideally, that's correct. This PR does not implement the logic to detect this.

Comment (VincentLanglet) [reply tests/PHPStan/Analyser/nsrt/array-find-key.php:47]: What's the current behavior ? I assume the type is considered as mixed.

This means that merging the PR as if will introduce a lot of false positif PHPstan errors on level 7 (unions) when it's currently on level 9 (mixed).
That's why IMHO if a dynamic return type extension is introduced, the implementation need to handle such case.

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/array-find-key.php:47]: > What's the current behavior ? I assume the type is considered as mixed.

just copy the test over into the playground:
https://phpstan.org/r/ece55e5c-33bf-47cb-b538-35708d025dfa

File: tests/PHPStan/Analyser/nsrt/array-find.php (hunk 1 above)
Comment (zonuexe) [reply tests/PHPStan/Analyser/nsrt/array-find.php]: Fixed https://github.com/phpstan/phpstan-src/pull/3518/commits/f0d9c2578802baeb0e6b8ba8cdf8663c54487df3

File: tests/PHPStan/Analyser/nsrt/array-find-key.php (hunk 4 above)
Comment (zonuexe) [reply tests/PHPStan/Analyser/nsrt/array-find-key.php:32]: It seems like it can be improved at the same time as `array_filter()`.

https://phpstan.org/r/ab62bff1-d485-46ae-9dfa-1931ce320a25

File: tests/PHPStan/Analyser/nsrt/array-find.php (hunk 2 above)
Comment (zonuexe) [reply tests/PHPStan/Analyser/nsrt/array-find.php:38]: It seems like it can be improved at the same time as `array_filter()`.

https://phpstan.org/r/ab62bff1-d485-46ae-9dfa-1931ce320a25

File: tests/PHPStan/Analyser/nsrt/array-find-key.php (hunk 4 above)
Comment (zonuexe) [reply tests/PHPStan/Analyser/nsrt/array-find-key.php:37]: We can add a test case, but if you pass a `non-empty-array<mixed>` the function may return any `int|string|null`.

https://3v4l.org/Biok8

Comment (VincentLanglet) [reply tests/PHPStan/Analyser/nsrt/array-find-key.php:37]: Indeed, my bad.

File: src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php [hunk 5]
```diff
@@ -62,290 +27,7 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
...
-		}
-
-		return new Name($funcName);
-	}
-
-	/**
-	 * @param self::USE_* $mode
-	 * @return array{list<Arg>, ?Variable, ?Variable}
-	 */
-	private function createDummyArgs(int $mode): array
```
Comment (ondrejmirtes) [inline src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php]: Do this refactoring in a separate PR first please.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php (hunk 5 above)
Comment (zonuexe) [reply src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php]: Ok, create a pull request.

Comment (zonuexe) [reply src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php]: Cherry-picked https://github.com/phpstan/phpstan-src/pull/3606/

**Comment by zonuexe** [comment]
I just cherry-picked them, so there should be no conflicts when https://github.com/phpstan/phpstan-src/pull/3606 is merged.

**Comment by ondrejmirtes** [comment]
Fix CS please.

**Comment by zonuexe** [comment]
@ondrejmirtes I fixed and rebased.

File: src/Reflection/ParametersAcceptorSelector.php [hunk 6]
```diff
@@ -257,6 +258,70 @@ public static function selectFromArgs(
...
+						$acceptor->getResolvedTemplateTypeMap(),
+						$parameters,
+						$acceptor->isVariadic(),
+						$acceptor->getReturnType(),
+						$acceptor instanceof ParametersAcceptorWithPhpDocs ? $acceptor->getCallSiteVarianceMap() : TemplateTypeVarianceMap::createEmpty(),
+					),
+				];
+			}
+
+			if (isset($args[0]) && (bool) $args[0]->getAttribute(ArrayWalkArgVisitor::ATTRIBUTE_NAME)) {
```
Comment (ondrejmirtes) [inline src/Reflection/ParametersAcceptorSelector.php]: Why is there a section about array_walk and it has no tests? It should be a separate PR in any case.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Reflection/ParametersAcceptorSelector.php (hunk 6 above)
Comment (zonuexe) [reply src/Reflection/ParametersAcceptorSelector.php]: Oops, I copied too much existing code.
I'm sure you're right, I'll split out the changes to ParametersAcceptorSelector into a separate PR.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Now you’ve removed too much 😀 It’s useful to declare the callable parameters for two reasons:

1) To infer the parameter type inside the callable based on array (even when the parameter does not have a native type).
2) To report wrong callable parameter tpe when it’s different to the passed array.

Please add tests for these.

**Comment by zonuexe** [comment]
Ok, I'll get to it right away.

**Comment by ondrejmirtes** [comment]
Perfect, thank you!
